### PR TITLE
Tmva cross validation

### PR DESCRIPTION
**Pull request by ashlaban**
Expanding cross validation capabilities of TMVA
 - Changes default split from using bootstapping to using k-folds with a user-defined splitting criterion.
 - Generate ROOT files that can be analysed with TMVA Gui, both per-fold output and aggregate is supported.
 - Cross validation in application. Using the user-defined split function, a set of methods generated with CV can be used in the application phase with new data. This is sometimes referred to as cross evaluation.
 - Add tutorial for cross evaluation

### CODE REVIEW

**Comment by ashlaban** [comment]
The test failure of centos7/gcc49 is because that platform is lacking a BLAS installation and cannot run DNN. Should I remove the DNN from the test?

File: tmva/tmva/CMakeLists.txt [hunk 1]
```diff
@@ -16,7 +16,8 @@ set(headers1 Configurable.h Factory.h  MethodBase.h MethodCompositeBase.h
 	     MethodKNN.h MethodCFMlpANN.h MethodCFMlpANN_Utils.h MethodLikelihood.h
 	     MethodHMatrix.h MethodPDERS.h MethodBDT.h MethodDT.h MethodSVM.h MethodBayesClassifier.h
 	     MethodFDA.h MethodMLP.h MethodBoost.h
-	     MethodPDEFoam.h MethodLD.h MethodCategory.h MethodDNN.h)
+	     MethodPDEFoam.h MethodLD.h MethodCategory.h MethodDNN.h
+         MethodCrossValidation.h)
```
Comment (amadio) [inline tmva/tmva/CMakeLists.txt]: The formatting seems to be off here.

File: tmva/tmva/inc/TMVA/CrossValidation.h [hunk 2]
```diff
@@ -1,31 +1,35 @@
 // @(#)root/tmva $Id$
-// Author: Omar Zapata, Thomas James Stevenson and Pourya Vakilipourtakalou. 2016
-
-#ifndef ROOT_TMVA_CrossValidation
-#define ROOT_TMVA_CrossValidation
+// Author: Omar Zapata, Thomas James Stevenson and Pourya Vakilipourtakalou
+// Modified: Kim Albertsson 2017
```
Comment (amadio) [inline tmva/tmva/inc/TMVA/CrossValidation.h]: 2017 -> 2018

File: tmva/tmva/inc/TMVA/CrossValidation.h [hunk 3]
```diff
@@ -1,31 +1,35 @@
...
+
+/**********************************************************************************
+ * Project: TMVA - a Root-integrated toolkit for multivariate data analysis       *
+ * Package: TMVA                                                                  *
+ * Class  : Factory                                                               *
+ *                                                                                *
+ * Copyright (c) 2016-2018:                                                       *
+ *      CERN, Switzerland                                                         *
```
Comment (amadio) [inline tmva/tmva/inc/TMVA/CrossValidation.h]: Long story, but copyright must be to Rene Brun and Fons Rademakers.

File: tmva/tmva/inc/TMVA/CvSplit.h [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+ * Description:                                                                   *
+ *                                                                                *
+ * Authors (alphabetical):                                                        *
+ *      Kim Albertsson <kim.albertsson@cern.ch> - CERN/LTU                        *
+ *      Sergei Gleyzer <sergei.gleyzer@cern.ch> - CERN, Switzerland               *
+ *      Lorenzo Moneta <Lorenzo.Moneta@cern.ch> - CERN, Switzerland               *
+ *                                                                                *
+ * Copyright (c) 2005-2017:                                                       *
+ *      CERN, Switzerland                                                         *
+ *      ITM/UdeA, Colombia                                                        *
```
Comment (amadio) [inline tmva/tmva/inc/TMVA/CvSplit.h]: Same as above, copyright to Rene and Fons (same for other files as well).

Comment (lmoneta) [reply tmva/tmva/inc/TMVA/CvSplit.h]: @amadio, this comment is not valid! Please ignore it

**Comment by lmoneta** [comment]
Hi Kim, 

 The test should be excluded only when the DNNCPU is not defined, as it is done for the DNN tests.

I think there is no need to split this PR. I will keep as it is. 

Lorenzo

File: tmva/tmva/inc/TMVA/CvSplit.h (hunk 4 above)
Comment (amadio) [reply tmva/tmva/inc/TMVA/CvSplit.h]: Is it an exception for TMVA? For ROOT in general I was told that copyright must be assigned to Rene and Fons.

**Comment by ashlaban** [comment]
@phsft-bot build please!

**Comment by amadio** [comment]
Problems above are probably due to externals. Rebuilding should fix. Let's try.

**Comment by amadio** [comment]
@phsft-bot build

**Comment by lmoneta** [comment]
Hi Kim,

 The failures seem to be unrelated to the PR. Can you Kim please fix the conflict so the PR can be merged ? 

 Thanks

 Lorenzo

**Comment by ashlaban** [comment]
Done!

Lorenzo Moneta wrote:

**Comment by lmoneta** [comment]
Thank you Kim, I have now merged the PR

**Pull request merged**